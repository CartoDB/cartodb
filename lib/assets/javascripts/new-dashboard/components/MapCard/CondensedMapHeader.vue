<template>
  <div class="map-list-row">
    <div class="viz-column--main-info">
      <div class="cell cell--start">
      </div>
      <div class="cell cell--main" @click="changeOrder('name')">
        <span class="text is-small is-txtSoftGrey"
              :class="{
                'is-active': isOrderApplied('name'),
                'is-reversed': isReverseOrderApplied('name'),
                'element-sort': isSortable
              }">
          {{ $t(`MapListHeader.name`) }}
        </span>
      </div>
    </div>

    <div class="viz-column--extra-info">
      <div class="viz-column--statusL">
        <div class="cell cell--medium u-p--0">
          <span class="text is-small is-txtSoftGrey"
                :class="{
                  'element-sort': false
                }">
            {{ $t(`MapListHeader.type`) }}
          </span>
        </div>
        <div class="cell cell--large" @click="changeOrder('updated_at')">
          <span class="text is-small is-txtSoftGrey"
                :class="{
                  'is-active': isOrderApplied('updated_at'),
                  'is-reversed': isReverseOrderApplied('updated_at'),
                  'element-sort': isSortable
                }">
            {{ $t(`MapListHeader.lastModified`) }}
          </span>
        </div>
        <div class="cell cell--xsmall u-txt-right u-p--0" @click="changeOrder('mapviews')" v-if="showViews">
          <span class="text is-small is-txtSoftGrey"
                :class="{
                  'is-active': isOrderApplied('mapviews'),
                  'is-reversed': isReverseOrderApplied('mapviews'),
                  'element-sort': isSortable
                }">
            {{ $t(`MapListHeader.views`) }}
          </span>
        </div>
      </div>

      <div class="viz-column--shareS">
        <div class="cell cell--small" @click="changeOrder('privacy')">
          <span class="text is-small is-txtSoftGrey"
                :class="{
                  'is-active': isOrderApplied('privacy'),
                  'is-reversed': isReverseOrderApplied('privacy'),
                  'element-sort': isSortable
                }">
            {{ $t(`MapListHeader.privacy`) }}
          </span>
        </div>
        <div class="cell cell--end"></div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'CondensedMapHeader',
  props: {
    order: String,
    orderDirection: String,
    isSortable: {
      type: Boolean,
      default: true
    },
    showViews: {
      type: Boolean,
      default: true
    }
  },
  methods: {
    changeOrder (order) {
      if (this.order === order) {
        return this.setOrder(
          order,
          this.getOppositeOrderDirection(this.orderDirection)
        );
      }

      this.setOrder(order);
    },

    getOppositeOrderDirection (order) {
      if (order === 'desc') {
        return 'asc';
      }

      if (order === 'asc') {
        return 'desc';
      }
    },

    isOrderApplied (order) {
      return order === this.order;
    },

    isReverseOrderApplied (order) {
      return order === this.order && this.orderDirection === 'asc';
    },

    setOrder (order, direction = 'desc') {
      this.$emit('orderChanged', { order, direction });
    }
  }
};
</script>

<style scoped lang="scss">
@import 'new-dashboard/styles/variables';

.map-list-row {
  display: flex;
  align-items: center;
  width: 100%;
  margin-bottom: 1px;
  padding: 14px;
  overflow: hidden;
  border-bottom: 1px solid $softblue;
  background-color: $white;
  cursor: default;
}

.cell--start {
  display: flex;
  align-items: center;
  align-self: flex-start;
  width: 58px;
  height: 100%;
}

.cell--end {
  width: 38px;
}

.element-sort {
  &.is-active,
  &:hover {
    color: $text__color;
    cursor: pointer;

    &::after {
      content: '';
      position: absolute;
      width: 14px;
      height: 100%;
      margin-left: 4px;
      transition: all 0.25s cubic-bezier(0.4, 0.01, 0.165, 0.99);
      background-image: url('../../assets/icons/datasets/chevron.svg');
      background-repeat: no-repeat;
      background-position: center;
    }

    &.is-reversed {
      &::after {
        transform: rotate(180deg);
      }
    }
  }
}
</style>
