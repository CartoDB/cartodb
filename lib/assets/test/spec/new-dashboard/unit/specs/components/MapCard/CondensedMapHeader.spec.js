import { shallowMount } from '@vue/test-utils';
import CondensedMapHeader from 'new-dashboard/components/MapCard/CondensedMapHeader';

const $t = key => key;

describe('CondensedMapHeader.vue', () => {
  let condensedMapHeader;
  beforeEach(() => {
    condensedMapHeader = createCondensedMapHeader();
  });

  it('should render correct contents', () => {
    expect(condensedMapHeader).toMatchSnapshot();
  });

  describe('Methods', () => {
    describe('changeOrder', () => {
      it('should call setOrder with passed order', () => {
        const setOrder = jest.fn();

        const header = createCondensedMapHeader({
          methods: { setOrder }
        });

        header.vm.changeOrder('name');

        expect(setOrder).toHaveBeenCalledWith('name');
      });

      it('should call setOrder with opposite order if passed order was applied already', () => {
        const setOrder = jest.fn();

        const header = createCondensedMapHeader({
          methods: { setOrder }
        });

        header.vm.changeOrder('updated_at');

        expect(setOrder).toHaveBeenCalledWith('updated_at', 'asc');
      });
    });

    describe('getOppositeOrderDirection', () => {
      it('should return asc when passing desc', () => {
        expect(condensedMapHeader.vm.getOppositeOrderDirection('asc')).toBe('desc');
      });

      it('should return desc when passing asc', () => {
        expect(condensedMapHeader.vm.getOppositeOrderDirection('desc')).toBe('asc');
      });
    });

    describe('isOrderApplied', () => {
      it('should return true if order is applied', () => {
        expect(condensedMapHeader.vm.isOrderApplied('updated_at')).toBe(true);
      });
    });

    describe('isReversedOrderApplied', () => {
      it('should return true if reverse order is applied', () => {
        condensedMapHeader.setData({
          order: 'updated_at',
          orderDirection: 'asc'
        });
        expect(condensedMapHeader.vm.isReverseOrderApplied('updated_at')).toBe(true);
      });
    });

    describe('setOrder', () => {
      it("should emit 'changeOrder' event with order and direction", () => {
        condensedMapHeader.vm.setOrder('name', 'asc');
        expect(condensedMapHeader.emitted('orderChanged')[0]).toEqual([
          { direction: 'asc', order: 'name' }
        ]);
      });
    });
  });
});

function createCondensedMapHeader (overridenComponentOptions) {
  return shallowMount(CondensedMapHeader, {
    mocks: {
      $t
    },
    propsData: {
      order: 'updated_at',
      orderDirection: 'desc'
    },
    ...overridenComponentOptions
  });
}
