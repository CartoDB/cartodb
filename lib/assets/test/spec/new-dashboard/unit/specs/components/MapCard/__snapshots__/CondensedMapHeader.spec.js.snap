// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CondensedMapHeader.vue should render correct contents 1`] = `
<div class="map-list-row">
  <div class="viz-column--main-info">
    <div class="cell cell--start"></div>
    <div class="cell cell--main"><span class="text is-small is-txtSoftGrey element-sort">
        MapListHeader.name
      </span></div>
  </div>
  <div class="viz-column--extra-info">
    <div class="viz-column--statusL">
      <div class="cell cell--medium u-p--0"><span class="text is-small is-txtSoftGrey">
          MapListHeader.type
        </span></div>
      <div class="cell cell--large"><span class="text is-small is-txtSoftGrey is-active element-sort">
          MapListHeader.lastModified
        </span></div>
      <div class="cell cell--xsmall u-txt-right u-p--0"><span class="text is-small is-txtSoftGrey element-sort">
          MapListHeader.views
        </span></div>
    </div>
    <div class="viz-column--shareS">
      <div class="cell cell--small"><span class="text is-small is-txtSoftGrey element-sort">
          MapListHeader.privacy
        </span></div>
      <div class="cell cell--end"></div>
    </div>
  </div>
</div>
`;
